### PR TITLE
fix type annotations for arguments in training_args

### LINKED ISSUE #24538: Incorrect typing of `fsdp`, `fsdp_config`, and `sharded_ddp` in `TrainingArguments`
### System Info

- `transformers` version: 4.29.2
- Platform: macOS-14.0-arm64-arm-64bit
- Python version: 3.10.6
- Huggingface_hub version: 0.15.1
- Safetensors version: not installed
- PyTorch version (GPU?): 1.13.0.dev20220902 (False)
- Tensorflow version (GPU?): 2.9.0 (True)
- Flax version (CPU?/GPU?/TPU?): not installed (NA)
- Jax version: not installed
- JaxLib version: not installed
- Using GPU in script?: No
- Using distributed or parallel set-up in script?: No

### Who can help?

@sgugger 

### Information

- [ ] The official example scripts
- [X] My own modified scripts

### Tasks

- [ ] An officially supported task in the `examples` folder (such as GLUE/SQuAD, ...)
- [X] My own task or dataset (give details below)

### Reproduction

When initializing a `pydantic.BaseModel` as follows:
```python
from pydantic import BaseModel
from transformers.training_args import TrainingArguments

class MyTrainingArguments(TrainingArguments):
    def __init__(self, **kwargs):
        super().__init__(**kwargs)
        self.my_arg = "my_arg"


class MyModel(BaseModel):
    training_args: MyTrainingArguments


model = MyModel(training_args=MyTrainingArguments(output_dir=""))
```
The following `ValidationErrors` occur:
```shell
ValidationError: 4 validation errors for MyModel
training_args -> debug
  str type expected (type=type_error.str)
training_args -> sharded_ddp
  str type expected (type=type_error.str)
training_args -> fsdp
  str type expected (type=type_error.str)
training_args -> fsdp_config
  str type expected (type=type_error.str)
```

Since `debug` has been fixed in #24033, my main concern are the others.
After investigation, I discovered that the `__post_init__()`-method changes these parameters from their default `str` values to for example `dict`, `bool`, or `List`.  This becomes a problem for Pydantic (and other type-checkers) since the validation will be incorrect, while the docstring of `TrainingArguments` describes the following for these parameters:
```python
"""
sharded_ddp (`bool`, `str` or list of [`~trainer_utils.ShardedDDPOption`], *optional*, defaults to `False`)
fsdp (`bool`, `str` or list of [`~trainer_utils.FSDPOption`], *optional*, defaults to `False`)
fsdp_config (`str` or `dict`, *optional*)
"""
```

### Expected behavior

I would like to resolve these issues by providing the correct typehinting. This could look as follows:
```python
sharded_ddp: Union[Optional[str], bool, List[ShardedDDPOption]]
fsdp: Union[Optional[str], bool, List[FSDPOption]]
fsdp_config: Union[Optional[str], Dict]
```
I checked this configuration and it resolves the issue.

### PR DESCRIPTION
**Pull request by shauray8**
# What does this PR do?

Fixes #24538 which is basically fixing type annotations for `fsdp`, `fsdp_config`, and `sharded_ddp` in training_args.py

## Who can review?
maybe @sgugger

### CODE REVIEW

File: src/transformers/training_args.py [hunk 1]
```diff
@@ -944,7 +944,7 @@ class TrainingArguments:
             )
         },
     )
-    sharded_ddp: str = field(
+    sharded_ddp: Union[Optional[str], bool, List[ShardedDDPOption]] = field(
```
Comment (sgugger) [inline src/transformers/training_args.py:947]: This is wrong, the Optional should come before the Union

File: src/transformers/training_args.py [hunk 2]
```diff
@@ -955,7 +955,7 @@ class TrainingArguments:
             ),
         },
     )
-    fsdp: str = field(
+    fsdp: Union[Optional[str], bool, List[FSDPOption]] = field(
```
Comment (sgugger) [inline src/transformers/training_args.py:958]: Same here

File: src/transformers/training_args.py [hunk 3]
```diff
@@ -976,7 +976,7 @@ class TrainingArguments:
             )
         },
     )
-    fsdp_config: Optional[str] = field(
+    fsdp_config: Union[Optional[str], Dict] = field(
```
Comment (sgugger) [inline src/transformers/training_args.py:979]: This was already taken care of in #24549

**Review by sgugger — COMMENTED** [review]
Thanks for your PR, I left a few comments.

File: src/transformers/training_args.py [hunk 4]
```diff
@@ -944,8 +944,8 @@ class TrainingArguments:
             )
         },
     )
-    sharded_ddp: str = field(
-        default="",
+    sharded_ddp: Optional[Union[str, bool, List[ShardedDDPOption]]] = field(
+        default=False,
```
Comment (sgugger) [inline src/transformers/training_args.py:948]: The default shouldn't be touched.

File: src/transformers/training_args.py [hunk 5]
```diff
@@ -955,8 +955,8 @@ class TrainingArguments:
             ),
         },
     )
-    fsdp: str = field(
-        default="",
+    fsdp: Optional[Union[str, bool, List[FSDPOption]]] = field(
+        default=False,
```
Comment (sgugger) [inline src/transformers/training_args.py:959]: Same here.

File: src/transformers/training_args.py (hunk 4 above)
Comment (shauray8) [reply src/transformers/training_args.py:948]: the documentation says that it defaults to False - sharded_ddp (`bool`, `str` or list of [`~trainer_utils.ShardedDDPOption`], *optional*, defaults to `False`). but if the default variable has to be " ", I'll try doing it some other way.

File: src/transformers/training_args.py (hunk 5 above)
Comment (shauray8) [reply src/transformers/training_args.py:959]: same for this - ```fsdp (`bool`, `str` or list of [`~trainer_utils.FSDPOption`], *optional*, defaults to `False`)```

File: src/transformers/training_args.py (hunk 4 above)
Comment (sgugger) [reply src/transformers/training_args.py:948]: Then let's fix the documentation as a change in default is potentially braking.

Comment (shauray8) [reply src/transformers/training_args.py:948]: Okay, changing default to `""` in the documentation as well as in the code.

**Comment by shauray8** [comment]
@sgugger I don't particularly understand why this error occurs in examples_flax - ``` argparse.ArgumentError: argument --sharded_ddp: invalid typing.Union[str, bool, typing.List[transformers.trainer_utils.ShardedDDPOption], NoneType] value: ' ' ```

**Comment by shauray8** [comment]
@sgugger `bool` breaks `--sharded_ddp`, I think we can still maintain Boolean arguments with string itself and 
https://github.com/huggingface/transformers/blob/20d6b84613984f2497587a62774704882ccbeee6/src/transformers/hf_argparser.py#L168-L173
with this `--sharded_ddp` and `--fsdp` defaults to string

**Review by sgugger — APPROVED** [review]
LGTM, thanks!

**Pull request merged**